src/Ledgers.jl

```diff
@@ -18,102 +18,88 @@ import Instruments: instrument, symbol, amount, name, currency
+account_type(::Union{AccountInfo{AT},AccountGroup{AT}}) where {AT} = AT
+code(acc::Union{<:AccountInfo,<:AccountGroup}) = acc.code
+name(acc::Union{<:AccountInfo,<:AccountGroup}) = acc.name
 
-struct Entry{D,C}
-    debit::D
-    credit::C
-end
+parent(group::AccountGroup) = group.parent
+subaccounts(group::AccountGroup) = group.subaccounts
+subgroups(group::AccountGroup) = group.subgroups
 
-Base.show(io::IO, e::Entry{<:Account,<:Account}) =
-    print(io, "Entry:\n", "  Debit: $(e.debit)\n", "  Credit: $(e.credit)")
+id(acc::Account) = acc.id
+id(info::AccountInfo) = id(account(info))
 
-abstract type AccountType end
-struct Credit <: AccountType end
-struct Debit <: AccountType end
+balance(acc::Account) = acc.balance
+balance(info::AccountInfo) = balance(account(info))
+balance(group::AccountGroup) = 
+    sum(map(info->balance(info), subaccounts(group))) + 
+    sum(map(grp->balance(grp), subgroups(group)))
 
-Base.show(io::IO, ::Type{Debit}) = print(io, "Debit")
-Base.show(io::IO, ::Type{Credit}) = print(io, "Credit")
+instrument(::Account{P}) where {P<:Position} = instrument(P)
+instrument(info::AccountInfo) = instrument(account(info))
 
-# SPJ: this does not maintain the distinction we'd talked about, of keeping
-# all debit accounts and all credit accounts in with separate types.
-# This will end up doing a lot of processing at run-time, and will be relatively
-# rather slow compared to what we had discussed.
+symbol(::Account{P}) where {P<:Position} = symbol(P)
+symbol(info::AccountInfo) = symbol(account(info))
 
-struct AccountInfo{T<:AccountType}
-    account::Account
-    name::String
-    parent::Union{Nothing,AccountInfo}
-    subaccounts::Vector{AccountInfo}
+currency(::Account{P}) where {P<:Position} = currency(P)
+currency(info::AccountInfo) = currency(account(info))
 
-    function AccountInfo(::Type{T}, account, name, parent=nothing) where {T<:AccountType}
-        ag = new{T}(account, name, parent, Vector{AccountInfo}())
-        (parent === nothing) || push!(parent.subaccounts, ag)
-        return ag
-    end
-end
+amount(acc::Account) = amount(balance(acc))
+amount(info::AccountInfo) = amount(balance(account(info)))
 
-account(info::AccountInfo) = info.account
-name(info::AccountInfo) = info.name
-parent(info::AccountInfo) = info.parent
-subaccounts(info::AccountInfo) = info.subaccounts
+debit!(acc::Account, amt::Position) = (acc.balance += amt)
+credit!(acc::Account, amt::Position) = (acc.balance -= amt)
 
-id(info::AccountInfo) = id(account(info))
-balance(info::AccountInfo{Debit}) = 
-    isempty(subaccounts(info)) ? balance(account(info)) : 
-    balance(account(info)) + sum(map(info->balance(account(info)), subaccounts(info)))
-balance(info::AccountInfo{Credit}) = 
-    isempty(subaccounts(info)) ? -balance(account(info)) :
-    -balance(account(info)) - sum(map(info->balance(account(info)), subaccounts(info)))
+struct Entry{D<:Account,C<:Account}
```

Comment:
Are D and C ever different types?

Reply:
Good question. Let's address this later.